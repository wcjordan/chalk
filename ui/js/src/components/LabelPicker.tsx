import React, { useCallback } from 'react';
import { StyleSheet, View, ViewStyle } from 'react-native';
import { Modal } from 'react-native-paper';
import LabelChip from './LabelChip';

interface Style {
  labelPickerView: ViewStyle;
  modalView: ViewStyle;
}

const styles = StyleSheet.create<Style>({
  labelPickerView: {
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
  modalView: {
    width: '80%',
    marginHorizontal: 'auto',
  },
});

const LabelPicker: React.FC<Props> = function (props: Props) {
  const { selectedLabels, setTodoLabelingId, updateTodoLabels } = props;

  const updateTodoLabelCb = useCallback(
    (label) => {
      const label_dict = Object.assign({}, selectedLabels, {
        [label]: !selectedLabels[label],
      });
      const label_set = Object.keys(label_dict).reduce(
        (acc: string[], label) => {
          if (label_dict[label]) {
            acc.push(label);
          }
          return acc;
        },
        [],
      );
      updateTodoLabels(label_set);
    },
    [updateTodoLabels, selectedLabels],
  );

  const dismissLabeling = useCallback(() => {
    setTodoLabelingId(null);
  }, [setTodoLabelingId]);

  const chips = props.labels.map((label) => (
    <LabelChip
      key={label}
      label={label}
      selected={selectedLabels[label] || false}
      updateTodoLabel={updateTodoLabelCb}
    />
  ));
  return (
    <Modal
      onDismiss={dismissLabeling}
      style={styles.modalView}
      visible={props.visible}
    >
      <View style={styles.labelPickerView}>{chips}</View>
    </Modal>
  );
};

type Props = {
  labels: string[];
  selectedLabels: { [label: string]: boolean };
  updateTodoLabels: (label_set: string[]) => void;
  setTodoLabelingId: (id: number | null) => void;
  visible: boolean;
};

export default LabelPicker;
