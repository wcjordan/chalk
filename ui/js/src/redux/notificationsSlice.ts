import { createSlice } from '@reduxjs/toolkit';

import { NotificationsState } from './types';

const initialState: NotificationsState = {
  notificationQueue: [],
};
export default createSlice({
  name: 'notifications',
  initialState,
  reducers: {
    addNotification: (state, action) => {
      state.notificationQueue.push(action.payload);
    },
    dismissNotification: (state) => {
      state.notificationQueue.shift();
    },
  },
});
