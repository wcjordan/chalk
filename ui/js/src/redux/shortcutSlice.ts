import { createSlice } from '@reduxjs/toolkit';

import { ShortcutState } from './types';

const initialState: ShortcutState = {
  operations: [],
  latestGeneration: 0,
};
export default createSlice({
  name: 'shortcuts',
  initialState,
  reducers: {
    addEditTodoOperation: (state, action) => {
      state.operations.push({
        type: 'EDIT_TODO',
        payload: action.payload,
        generation: state.latestGeneration,
      });
    },
    addMoveTodoOperation: (state, action) => {
      state.operations.push({
        type: 'MOVE_TODO',
        payload: action.payload,
        generation: state.latestGeneration,
      });
    },
    incrementGenerations: (state) => {
      state.latestGeneration = state.latestGeneration + 1;
    },
    clearOperationsUpThroughGeneration: (state, action) => {
      state.operations = state.operations.filter(
        (op) => op.generation > action.payload,
      );
    },
  },
});
