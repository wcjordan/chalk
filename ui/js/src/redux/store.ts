import { configureStore } from '@reduxjs/toolkit';
import rootReducer, { listTodos } from './reducers';

const store = configureStore({
  reducer: rootReducer,
});

// TODO dispose
window.setInterval(() => store.dispatch(listTodos()), 1000);

export default store;
