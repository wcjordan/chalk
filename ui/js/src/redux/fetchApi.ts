import Cookies from 'js-cookie';
import { Platform } from 'react-native';

import { getEnvFlags } from '../helpers';
import { RootState } from './store';

const webCsrfToken = Cookies.get('csrftoken') as string;
export function getCsrfToken(getState: () => RootState): string {
  if (Platform.OS === 'web') {
    return webCsrfToken;
  }
  return getState().workspace.csrfToken as string;
}

export async function list<T>(apiUri: string): Promise<Array<T>> {
  const response = await fetch(apiUri, getRequestOpts('GET'));
  return handleResponse<Array<T>>(response);
}

export async function patch<T, Q extends { id: number }>(
  apiUri: string,
  entryPatch: Q,
  csrfToken: string,
): Promise<T> {
  const requestOpts = getRequestOpts('PATCH', csrfToken);
  requestOpts.body = JSON.stringify(entryPatch);
  const response = await fetch(`${apiUri}${entryPatch.id}/`, requestOpts);
  return handleResponse<T>(response);
}

export async function create<T, Q>(
  apiUri: string,
  newEntry: Q,
  csrfToken: string,
): Promise<T> {
  return postRequest<T, Q>(apiUri, newEntry, csrfToken);
}

export async function postRequest<T, Q>(
  uri: string,
  data: Q,
  csrfToken: string,
): Promise<T> {
  const requestOpts = getRequestOpts('POST', csrfToken);
  requestOpts.body = JSON.stringify(data);
  const response = await fetch(uri, requestOpts);
  return handleResponse<T>(response);
}

function getRequestOpts(
  method: string,
  csrfToken: string | null = null,
): RequestInit {
  const headers: HeadersInit = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  };
  if (csrfToken) {
    headers['X-CSRFToken'] = csrfToken;
  }

  const credentials: RequestCredentials = Platform.select({
    native: 'include',
    default: 'same-origin',
  });

  return {
    credentials: credentials,
    headers: headers,
    method: method,
  };
}

async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    throw new Error(`${response.status} ${response.statusText}`);
  }
  return await response.json();
}

export function getWsRoot(): string {
  if (Platform.OS === 'web') {
    return '';
  }

  const subdomain =
    getEnvFlags().ENVIRONMENT === 'prod' ? 'chalk' : 'chalk-dev';
  return `http://${subdomain}.flipperkid.com/`
}

// Used to exchange login token for session cookie in mobile login flow
export async function completeAuthCallback(token: string) {
  return await fetch(
    `${getWsRoot()}api/todos/auth_callback/?code=${token}`,
    getRequestOpts('GET'),
  );
}

export async function logSessionData(
  dataString: string,
  csrfToken: string,
): Promise<string> {
  
  const requestOpts = getRequestOpts('POST', csrfToken);
  requestOpts.body = dataString;
  const response = await fetch(
    `${getWsRoot()}api/todos/log_session_data/`,
    requestOpts,
  );
  return handleResponse<string>(response);
}
