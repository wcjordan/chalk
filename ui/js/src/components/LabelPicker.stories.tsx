import React from 'react';
import { StyleSheet, View } from 'react-native';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { Provider } from 'react-redux';
import { setupStore } from '../redux/store';
import LabelPicker from './LabelPicker';

const labels = [
  'low-energy',
  'high-energy',
  'vague',
  'work',
  'home',
  'errand',
  'mobile',
  'desktop',
  'email',
  'urgent',
  '5 minutes',
  '25 minutes',
  '60 minutes',
];
const selectedLabels = {
  '5 minutes': true,
  work: true,
  home: true,
  'low-energy': true,
  mobile: true,
};

const styles = StyleSheet.create({
  wrapper: {
    paddingTop: '100px',
    padddingBottom: '100px',
  },
});
const wrapper = (component) => (
  <SafeAreaProvider>
    <Provider store={setupStore()}>
      <View style={styles.wrapper}>{component}</View>
    </Provider>
  </SafeAreaProvider>
);

export default {
  title: 'Label Picker',
  component: LabelPicker,
};
export const DefaultLabelPicker: React.FC = () =>
  wrapper(
    <LabelPicker
      labels={labels}
      selectedLabels={selectedLabels}
      visible={true}
    />,
  );
