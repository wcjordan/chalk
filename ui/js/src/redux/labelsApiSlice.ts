import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { ApiState, ReduxState, Label } from './types';
import { getWsRoot, list } from './fetchApi';

const API_NAME = 'labelsApi';

// Exported for testing
export function getLabelsApi(): string {
  return `${getWsRoot()}api/todos/labels/`;
}

const initialState: ApiState<Label> = {
  entries: [],
  initialLoad: true,
  loading: false,
};

export const listLabels = createAsyncThunk<
  Label[],
  void,
  { state: ReduxState }
>(`${API_NAME}/list`, async () => list<Label>(getLabelsApi()), {
  condition: (_unused, { getState }) => {
    return !getState().labelsApi.loading;
  },
});

export default createSlice({
  name: API_NAME,
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(listLabels.pending, (state) => {
        state.loading = true;
      })
      .addCase(listLabels.fulfilled, (state, action) => {
        state.initialLoad = false;
        state.loading = false;
        state.entries = action.payload;
      })
      .addCase(listLabels.rejected, (state, action) => {
        state.initialLoad = false;
        state.loading = false;
        console.warn(`Loading Labels failed. ${action.error.message}`);
      });
  },
});
