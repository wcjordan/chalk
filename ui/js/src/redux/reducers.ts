import { ThunkAction } from 'redux-thunk';
import { Action } from '@reduxjs/toolkit';

import { ReduxState, TodoPatch } from './types';
import labelsApiSlice, { listLabels } from './labelsApiSlice';
import notificationsSlice from './notificationsSlice';
import shortcutSlice from './shortcutSlice';
import todosApiSlice, {
  createTodo,
  listTodos,
  updateTodo as updateTodoApi,
} from './todosApiSlice';
import workspaceSlice from './workspaceSlice';
import { completeAuthCallback } from './fetchApi';

type AppThunk = ThunkAction<void, ReduxState, unknown, Action<string>>;

export const updateTodo =
  (todoPatch: TodoPatch): AppThunk =>
  (dispatch) => {
    return Promise.all([
      dispatch(
        notificationsSlice.actions.addNotification(
          `Saving Todo: ${todoPatch.description}`,
        ),
      ),
      dispatch(workspaceSlice.actions.setEditTodoId(null)),
      dispatch(shortcutSlice.actions.addEditTodoOperation(todoPatch)),
      dispatch(updateTodoApi(todoPatch)),
    ]);
  };

export const updateTodoLabels =
  (labels: string[]): AppThunk =>
  (dispatch, getState) => {
    const todoId = getState().workspace.labelTodoId;
    if (todoId === null) {
      throw new Error('Unable to edit a todo w/ null ID');
    }

    return dispatch(
      updateTodoApi({
        id: todoId,
        labels,
      }),
    );
  };

// Used to exchange login token for session cookie in mobile login flow
export const completeAuthentication =
  (accessToken: string): AppThunk =>
  async (dispatch) => {
    try {
      const response = await completeAuthCallback(accessToken);
      const cookies = response.headers.get('set-cookie')?.split(' ');
      const csrfToken = cookies
        ?.find((val) => val.startsWith('csrftoken'))
        ?.slice(10, -1);

      if (response.status !== 200) {
        const responseText = await response.text();
        const message = `Login failed with status: ${response.status} ${response.statusText}\n${responseText}`;
        dispatch(notificationsSlice.actions.addNotification(message));
        throw new Error(message);
      }

      if (!csrfToken) {
        dispatch(
          notificationsSlice.actions.addNotification(
            'Unexpectedly missing CSRFToken.  Please refresh and login again.',
          ),
        );
        throw new Error(
          `CSRFToken missing from set-cookie header\n${response.headers.get(
            'set-cookie',
          )}`,
        );
      }

      dispatch(
        workspaceSlice.actions.logIn({
          loggedIn: true,
          csrfToken: csrfToken,
        }),
      );
    } catch (ex) {
      dispatch(
        notificationsSlice.actions.addNotification(
          'Login failed with a network error.  Please refresh and login again.',
        ),
      );
      throw ex;
    }
  };

export const addNotification = notificationsSlice.actions.addNotification;
export const dismissNotification =
  notificationsSlice.actions.dismissNotification;
export const setEditTodoId = workspaceSlice.actions.setEditTodoId;
export const setLabelTodoId = workspaceSlice.actions.setLabelTodoId;
export const setWorkContext = workspaceSlice.actions.setWorkContext;
export const toggleLabel = workspaceSlice.actions.toggleLabel;
export const toggleShowCompletedTodos =
  workspaceSlice.actions.toggleShowCompletedTodos;
export const toggleShowLabelFilter =
  workspaceSlice.actions.toggleShowLabelFilter;
export { createTodo, listLabels, listTodos };
export default {
  labelsApi: labelsApiSlice.reducer,
  notifications: notificationsSlice.reducer,
  shortcuts: shortcutSlice.reducer,
  todosApi: todosApiSlice.reducer,
  workspace: workspaceSlice.reducer,
};
