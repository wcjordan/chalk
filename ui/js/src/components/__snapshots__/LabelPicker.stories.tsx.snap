// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label Picker DefaultLabelPicker smoke-test 1`] = `
<div class="css-175oi2r r-13awgt0">
  <div class="css-175oi2r r-fsv9o2 r-oejqat">
    <div aria-live="polite"
         class="css-175oi2r r-1p0dtai r-1d2f490 r-u8s1d r-zchlnj r-ipm5af r-105ug2t"
         data-testid="modal"
    >
      <button aria-label="Close modal"
              role="button"
              tabindex="0"
              class="css-175oi2r r-1loqt21 r-1otgn73 r-13awgt0"
              data-testid="modal-backdrop"
              type="button"
              style="background-color: rgba(50, 47, 55, 0.4); opacity: 1;"
      >
      </button>
      <div class="css-175oi2r r-1p0dtai r-1777fci r-1d2f490 r-u8s1d r-zchlnj r-ipm5af r-12vffkv"
           data-testid="modal-wrapper"
           style="margin-top: 0px; margin-bottom: 0px;"
      >
        <div class="css-175oi2r r-1niwhzg r-1777fci r-1jj8364 r-lchren r-6e0ovw r-105ug2t"
             data-testid="modal-surface"
             style="box-shadow: rgba(0, 0, 0, 0.3) 0px 1px 3px; opacity: 1;"
        >
          <div class="css-175oi2r r-hr69ez r-18u37iz r-1w6e6rj"
               data-testid="label-picker"
          >
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-1iizze8 r-1bq2mok r-105ug2t"
                 data-testid="chip-active-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip-active"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div class="css-175oi2r r-1kihuf0 r-tuq35u r-1m04atk r-9aemit">
                    <div dir="auto"
                         role="img"
                         tabindex="-1"
                         class="css-146c3p1 r-1niwhzg r-lrvibr r-633pao"
                         style="font-size: 18px; color: rgb(29, 25, 43); transform: scaleX(1); line-height: 18px; font-family: material-community; font-weight: normal; font-style: normal;"
                    >
                      󰄬
                    </div>
                  </div>
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    low-energy
                  </div>
                </div>
              </button>
            </div>
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
                 data-testid="chip-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    high-energy
                  </div>
                </div>
              </button>
            </div>
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
                 data-testid="chip-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    vague
                  </div>
                </div>
              </button>
            </div>
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-1iizze8 r-1bq2mok r-105ug2t"
                 data-testid="chip-active-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip-active"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div class="css-175oi2r r-1kihuf0 r-tuq35u r-1m04atk r-9aemit">
                    <div dir="auto"
                         role="img"
                         tabindex="-1"
                         class="css-146c3p1 r-1niwhzg r-lrvibr r-633pao"
                         style="font-size: 18px; color: rgb(29, 25, 43); transform: scaleX(1); line-height: 18px; font-family: material-community; font-weight: normal; font-style: normal;"
                    >
                      󰄬
                    </div>
                  </div>
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    work
                  </div>
                </div>
              </button>
            </div>
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-1iizze8 r-1bq2mok r-105ug2t"
                 data-testid="chip-active-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip-active"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div class="css-175oi2r r-1kihuf0 r-tuq35u r-1m04atk r-9aemit">
                    <div dir="auto"
                         role="img"
                         tabindex="-1"
                         class="css-146c3p1 r-1niwhzg r-lrvibr r-633pao"
                         style="font-size: 18px; color: rgb(29, 25, 43); transform: scaleX(1); line-height: 18px; font-family: material-community; font-weight: normal; font-style: normal;"
                    >
                      󰄬
                    </div>
                  </div>
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    home
                  </div>
                </div>
              </button>
            </div>
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
                 data-testid="chip-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    errand
                  </div>
                </div>
              </button>
            </div>
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-1iizze8 r-1bq2mok r-105ug2t"
                 data-testid="chip-active-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip-active"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div class="css-175oi2r r-1kihuf0 r-tuq35u r-1m04atk r-9aemit">
                    <div dir="auto"
                         role="img"
                         tabindex="-1"
                         class="css-146c3p1 r-1niwhzg r-lrvibr r-633pao"
                         style="font-size: 18px; color: rgb(29, 25, 43); transform: scaleX(1); line-height: 18px; font-family: material-community; font-weight: normal; font-style: normal;"
                    >
                      󰄬
                    </div>
                  </div>
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    mobile
                  </div>
                </div>
              </button>
            </div>
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
                 data-testid="chip-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    desktop
                  </div>
                </div>
              </button>
            </div>
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
                 data-testid="chip-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    email
                  </div>
                </div>
              </button>
            </div>
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
                 data-testid="chip-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    urgent
                  </div>
                </div>
              </button>
            </div>
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-1iizze8 r-1bq2mok r-105ug2t"
                 data-testid="chip-active-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip-active"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div class="css-175oi2r r-1kihuf0 r-tuq35u r-1m04atk r-9aemit">
                    <div dir="auto"
                         role="img"
                         tabindex="-1"
                         class="css-146c3p1 r-1niwhzg r-lrvibr r-633pao"
                         style="font-size: 18px; color: rgb(29, 25, 43); transform: scaleX(1); line-height: 18px; font-family: material-community; font-weight: normal; font-style: normal;"
                    >
                      󰄬
                    </div>
                  </div>
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    5 minutes
                  </div>
                </div>
              </button>
            </div>
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
                 data-testid="chip-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    25 minutes
                  </div>
                </div>
              </button>
            </div>
            <div class="css-175oi2r r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
                 data-testid="chip-container"
                 style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
            >
              <button role="button"
                      tabindex="0"
                      class="css-175oi2r r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
                      data-testid="chip"
                      type="button"
                      style="border-radius: 8px;"
              >
                <div class="css-175oi2r r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
                     style="padding-right: 0px;"
                >
                  <div dir="auto"
                       class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
                       style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
                  >
                    60 minutes
                  </div>
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
